src/append.jl

```diff
@@ -0,0 +1,168 @@
+  * `ntasks::Int`: number of concurrent threaded tasks to allow while writing input partitions out as arrow record batches; default is no limit; to disable multithreaded writing, pass `ntasks=1`
+"""
+function append end
+
+append(file::String; kw...) = x -> append(file::String, x; kw...)
+
+function append(file::String, tbl;
+        largelists::Bool=false,
+        compress::Union{Nothing, Symbol, LZ4FrameCompressor, ZstdCompressor}=nothing,
+        denseunions::Bool=true,
+        dictencode::Bool=false,
+        dictencodenested::Bool=false,
+        alignment::Int=8,
+        maxdepth::Int=DEFAULT_MAX_DEPTH,
+        ntasks=Inf)
+    if ntasks < 1
+        throw(ArgumentError("ntasks keyword argument must be > 0; pass `ntasks=1` to disable multithreaded writing"))
+    end
+    if compress === :lz4
+        compress = LZ4_FRAME_COMPRESSOR
+    elseif compress === :zstd
+        compress = ZSTD_COMPRESSOR
+    elseif compress isa Symbol
+        throw(ArgumentError("unsupported compress keyword argument value: $compress. Valid values include `:lz4` or `:zstd`"))
+    end
+
+    open(file, "r+") do io
+        append(io, tbl, largelists, compress, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    end
+
+    return file
+end
+
+function append(io, source, largelists, compress, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    bytes = Mmap.mmap(io)
+    stream = Arrow.Stream(bytes)
+    arrow_schema = Tables.schema(Tables.columns(stream))
+    dictencodings = stream.dictencodings
+    firstcols = toarrowtable(stream, dictencodings, largelists, compress, denseunions, dictencode, dictencodenested, maxdepth)
+    sch = Ref{Tables.Schema}(Tables.schema(firstcols))
+
+    # position file at the last footer
+    offset = footer_offset(bytes)
+    seek(io, length(bytes) - offset)
+
+    # TODO: we're probably not threadsafe if user passes own single compressor instance + ntasks > 1
+    # if ntasks > 1 && compres !== nothing && !(compress isa Vector)
+    #     compress = Threads.resize_nthreads!([compress])
+    # end
+    msgs = OrderedChannel{Message}(ntasks)
+    # build messages
+    blocks = (Block[], Block[])
+    # start message writing from channel
+    threaded = ntasks > 1
+    tsk = threaded ? (Threads.@spawn for msg in msgs
+        Base.write(io, msg, blocks, sch, alignment)
+    end) : (@async for msg in msgs
+        Base.write(io, msg, blocks, sch, alignment)
+    end)
+    anyerror = Threads.Atomic{Bool}(false)
+    errorref = Ref{Any}()
+    @sync for (i, tbl) in enumerate(Tables.partitions(source))
+        if anyerror[]
+            @error "error writing arrow data on partition = $(errorref[][3])" exception=(errorref[][1], errorref[][2])
+            error("fatal error writing arrow data")
+        end
+        @debug 1 "processing table partition i = $i"
+        tbl_schema = Tables.schema(tbl)
```

Comment:
Technically, we need to call `Tables.schema` on the result of `Tables.columns(tbl)`, though often they produce the same result. But we should probably pass the `stream` schema down to `process_partition` to do the comparison there.

Reply:
changed in https://github.com/JuliaData/Arrow.jl/pull/160/commits/7aca33da479fad8e98d1ff192d98adf9aba6e8b7 now

Reply:
Ok, I think this all looks good now; this is the only remaining piece. Specifically, we're calling `Tables.schema(tbl)`, where `tbl` is a partition iteration element from `Tables.partitions`. Technically, we need to call `Tables.schema(Tables.columns(tbl))`. We don't end up calling `Tables.columns` until `process_partition`. Probably easiest is to change `process_partition` to expect the 1st argument to already be the result of `Tables.columns(tbl)` and we can call it like that here:
```julia
cols = Tables.columns(tbl)
tbl_schema = Tables.schema(cols)
...
process_partition(cols, ...)
```
Then we just need to make sure we update the usage of `process_partition` in write.jl as well.

Reply:
done in https://github.com/JuliaData/Arrow.jl/pull/160/commits/0f0ab4bbfc2f72dc7e1ac41071a84465f7c58c01